README.txt
=====================
cp.recipe.cmd package
=====================

.. contents::

What is cp.recipe.cmd ?
=======================


This recipe is used to run one or more command lines.

I stole this from iw.recipe.cmd (http://pypi.python.org/pypi/iw.recipe.cmd/0.1)

It works differently tho, when it comes to executing shell commands.  iw.recipe.cmd would push each command out separately in it's own shell.  Here I push them out to a shell script, and then run the shell script.  This way things like CD and other things that require state within the shell work great.

options:
shell= Set a shell to execute (default /bin/sh) (do not put the #!)
on_install= whether to run or not on_install (defaults True)
on_update= whether to run or not on_update (defaults True)
install_cmd = Commands to run when install happens.
update_cmd = Commands to run when an update happens.

Also, I changed the way it works in the config file (example):

[commandexample]
recipe = cp.recipe.cmd
shell = /bin/sh
install_cmd =
   echo "install commands go here"
	cd /tmp
	echo `pwd`
	echo 'see, I exist in one shell instance.'
update_cmd =
	echo "update commands go here"
	

On install, install_cmd will be turned into a shell script, and then ran.
on update, update_cmd will be turned into a shell script and then ran.  If you want update_cmds to be the same you can do something like this:
update_cmd = ${commandexample:install_cmd}

(where commandexample is the name of your part)


I've added a new option shell=
You can put whateer you want in there, sme nice examples:
/usr/bin/env python
(to run the python interpreter)
the default is /bin/sh, so all your old code using this will work just fine.

Also, I check the output of the script and if something returns >0 then
CmdExecutionFailed is raised, along with the output of the error, and the path
of where the script is, so you can see what was generated, do debugging, and
fix it.


a python example using the new shell= is here:


[buildout]
parts = cmds

[cmds]
recipe = cp.recipe.cmd
shell=/usr/bin/env python
update_cmd =
	f = open('testfile'.'w')
	f.write('this is a testfile')
	f.close()



Below are the original docs (which can be ignored probably):
=======================



We need a config file::

  >>> cfg = """
  ... [buildout]
  ... parts = cmds
  ...
  ... [cmds]
  ... recipe = iw.recipe.cmd
  ... on_install=true
  ... cmds= %s
  ... """

  >>> test_file = join(sample_buildout, 'test.txt')
  >>> cmds = 'touch %s' % test_file
  >>> write(sample_buildout, 'buildout.cfg', cfg % cmds)

Ok, so now we can touch a file for testing::

  >>> print system(buildout)
  Installing cmds.

  >>> 'test.txt' in os.listdir(sample_buildout)
  True

And remove it::

  >>> test_file = join(sample_buildout, 'test.txt')
  >>> cmds = 'rm -f %s' % test_file
  >>> write(sample_buildout, 'buildout.cfg', cfg % cmds)

  >>> print system(buildout)
  Uninstalling cmds.
  Installing cmds.

  >>> 'test.txt' in os.listdir(sample_buildout)
  False

We can run more than one commands::

  >>> cmds = '''
  ... touch %s
  ... rm -f %s
  ... ''' % (test_file, test_file)

  >>> test_file = join(sample_buildout, 'test.txt')
  >>> cmds = 'rm -f %s' % test_file
  >>> write(sample_buildout, 'buildout.cfg', cfg % cmds)

  >>> print system(buildout)
  Updating cmds.

  >>> 'test.txt' in os.listdir(sample_buildout)
  False

We can also run some python code::

  >>> cfg = """
  ... [buildout]
  ... parts = py py2
  ...
  ... [py]
  ... recipe = iw.recipe.cmd:py
  ... on_install=true
  ... cmds= 
  ...   >>> sample_buildout = buildout.get('directory', '.')
  ...   >>> print os.listdir(sample_buildout)
  ...   >>> shutil.rmtree(os.path.join(sample_buildout, "bin"))
  ...   >>> print os.listdir(sample_buildout)
  ... [py2]
  ... recipe = iw.recipe.cmd:py
  ... on_install=true
  ... cmds=
  ...   >>> def myfunc(value):
  ...   ...     return value and True or False
  ...   >>> v = 20
  ...   >>> print myfunc(v)
  ... """

  >>> write(sample_buildout, 'buildout.cfg', cfg)

Ok, so now we run it::

  >>> print system(buildout)
  Uninstalling cmds.
  Installing py.
  ['.installed.cfg', 'bin', 'buildout.cfg', 'develop-eggs', 'eggs', 'parts']
  ['.installed.cfg', 'buildout.cfg', 'develop-eggs', 'eggs', 'parts']
  Installing py2.
  True




HISTORY.txt
Changelog
=========

0.4
----

- install hg plugin so HISTORY.txt makes it into distribution
  [sureshvv]

0.5 (unreleased)
----------------

- Nothing changed yet.


0.4 (2011-01-17)
----------------

- add HISTORY.txt to track release notes
  [sureshvv]

0.2 (2010-12-07)
-----------------

- added buildout test/example
  [csawyer]

- Implement the "shell" option
  [csawyer]

- Get exit status of command and raise exception when it fails.
  [csawyer]

0.1b (2009-2-2)
---------------

- updated so docs are accurate! update_cmd / install_cmd
  Thanks to kevin gill <kevin.gill@openapp.ie> for finding and fixing the bug!
  [csawyer]

0.1a:
-----------------------

- initial release
  [csawyer]


cp/recipe/cmd/tests/test_cmddocs.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 'Ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Generic Test case for 'cp.recipe.cmd' doctest
"""
__docformat__ = 'restructuredtext'

import zc.buildout.tests
import zc.buildout.testing

import unittest
import doctest
import sys
import os
from os.path import join

from zope.testing import doctest, renormalizing

def setUp(test):
    zc.buildout.testing.buildoutSetUp(test)
    zc.buildout.testing.install_develop('iw.recipe.template', test)

tearDown=zc.buildout.testing.buildoutTearDown

current_dir = os.path.dirname(__file__)

def doc_suite(test_dir, setUp=setUp, tearDown=tearDown, globs=None):
    """Returns a test suite, based on doctests found in /doctest."""
    suite = []
    if globs is None:
        globs = globals()

    globs['test_dir'] = current_dir

    flags = (doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE |
             doctest.REPORT_ONLY_FIRST_FAILURE)

    package_dir = os.path.split(test_dir)[0]
    if package_dir not in sys.path:
        sys.path.append(package_dir)

    doctest_dir = os.path.join(package_dir, 'docs')

    # filtering files on extension
    docs = [os.path.join(doctest_dir, doc) for doc in
            os.listdir(doctest_dir) if doc.endswith('.txt')]

    globs['test_dir'] = current_dir

    for test in docs:
        suite.append(doctest.DocFileSuite(test, optionflags=flags,
                                          globs=globs, setUp=setUp,
                                          tearDown=tearDown,
                                          module_relative=False,
                                          checker=renormalizing.RENormalizing([
                                                    zc.buildout.testing.normalize_path,
                                            ])))

    return unittest.TestSuite(suite)

def test_suite():
    """returns the test suite"""
    return doc_suite(current_dir)

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')



cp/recipe/cmd/__init__.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 'Ingeniweb'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Recipe cmd"""
import commands
import tempfile
import shutil
import os, sys


class CmdExecutionFailed(Exception):
	pass


class Cmd(object):
	"""This recipe is used by zc.buildout"""

	def __init__(self, buildout, name, options):
		self.buildout, self.name, self.options = buildout, name, options
		self.on_install = options.get('on-install', True)
		self.on_update = options.get('on-update', True)
		self.shell = options.get('shell','/bin/sh')

	def install(self):
		"""installer"""
		cmds = self.options.get('install-cmd', '')
		if self.on_install:
			self.execute(cmds)
		return tuple()

	def update(self):
		"""updater"""
		cmds = self.options.get('update-cmd', '')
		if self.on_update:
			self.execute(cmds)
		return tuple()

	def execute(self,cmds):
		"""run the commands
		"""
		cmds = cmds.strip()
		if not cmds:
			return
		if cmds:
			cmds = cmds.split('\n')
			dirname = tempfile.mkdtemp()
			lines = [line.strip() for line in cmds]
			tmpfile = os.path.join(dirname,'run.sh')
			fil = open(tmpfile,'w+')
			fil.write("#!%s\n" % self.shell)
			fil.write('\n'.join(lines))
			fil.close()
			#give execute permissions
			os.chmod(tmpfile,0700)
			status, results = commands.getstatusoutput(tmpfile)
			print results
			if status:
				raise CmdExecutionFailed,results
			return status


class Python(Cmd):

	def execute(self):
		"""run python code
		"""
		cmds = self.options.get('cmds', '')
		cmds = cmds.strip()
		def undoc(l):
			l = l.strip()
			l = l.replace('>>> ', '')
			l = l.replace('... ', '')
			return l

		if not cmds:
			return
		if cmds:
			name = self.name
			buildout = self.buildout
			options = self.options
			lines = cmds.split('\n')
			lines = [undoc(line) for line in lines if line.strip()]
			dirname = tempfile.mkdtemp()
			tmpfile = os.path.join(dirname, 'run.py')
			open(tmpfile, 'w').write('\n'.join(lines))
			execfile(tmpfile)
			shutil.rmtree(dirname)


